Userland/Applications/Mail/MailWidget.cpp

```diff
@@ -0,0 +1,484 @@
+        auto* potential_alternative_structure = look_for_alternative_body_structure(structure->data().get<IMAP::MultiPartBodyStructureData>(), position_stack);
+
+        if (potential_alternative_structure)
+            return potential_alternative_structure;
+
+        position_stack.take_last();
+        ++structure_index;
+    }
+
+    return nullptr;
+}
+
+Vector<MailWidget::Alternative> MailWidget::get_alternatives(IMAP::MultiPartBodyStructureData const& multi_part_body_structure_data) const
+{
+    Vector<int> position_stack;
+
+    auto* alternative_body_structure = look_for_alternative_body_structure(multi_part_body_structure_data, position_stack);
+    if (!alternative_body_structure)
+        return {};
+
+    Vector<MailWidget::Alternative> alternatives;
+    alternatives.ensure_capacity(alternative_body_structure->bodies.size());
+
+    int alternative_index = 1;
+    for (auto& alternative_body : alternative_body_structure->bodies) {
+        VERIFY(alternative_body->data().has<IMAP::BodyStructureData>());
+
+        position_stack.append(alternative_index);
+
+        MailWidget::Alternative alternative = {
+            .body_structure = alternative_body->data().get<IMAP::BodyStructureData>(),
+            .position = position_stack,
+        };
+        alternatives.append(alternative);
+
+        position_stack.take_last();
+        ++alternative_index;
+    }
+
+    return alternatives;
+}
+
+bool MailWidget::is_supported_alternative(Alternative const& alternative) const
+{
+    auto lowercase_type = alternative.body_structure.type.to_lowercase();
+    auto lowercase_subtype = alternative.body_structure.subtype.to_lowercase();
+    return lowercase_type == "text" && lowercase_subtype.is_one_of("plain", "html");
+}
+
+void MailWidget::selected_mailbox()
+{
+    const auto& index = m_mailbox_list->selection().first();
+
+    if (!index.is_valid())
+        return;
+
+    auto& base_node = *static_cast<BaseNode*>(index.internal_data());
+
+    if (is<AccountNode>(base_node)) {
+        // FIXME: Do something when clicking on an account node.
+        return;
+    }
+
+    auto& mailbox_node = verify_cast<MailboxNode>(base_node);
+    auto& mailbox = mailbox_node.mailbox();
+
+    VERIFY(!(mailbox.flags & (unsigned)IMAP::MailboxFlag::NoSelect));
+
+    auto response = m_imap_client->select(mailbox.name)->await().release_value();
+
+    if (response.status() != IMAP::ResponseStatus::OK) {
+        dbgln("Failed to select mailbox. The server says: '{}'", response.response_text());
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to select mailbox. The server says: '{}'", response.response_text()));
+        return;
+    }
+
+    Vector<IMAP::SearchKey> keys;
+    // clang-format off
+    keys.append(IMAP::SearchKey {
+        IMAP::SearchKey::All {}
+    });
+    // clang-format on
+    response = m_imap_client->search({}, move(keys), false)->await().release_value();
```

Comment:
This is unnecessary, since the mails will always be numbered from 1 upto the number of mails that exist (which is returned as part of the select response).

Reply:
Great, fixed.